Add randomness case to the autograd notes

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78106
* __->__ #78617

I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

lezcano committed:
Add randomness case to the autograd notes

I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

[ghstack-poisoned]

lezcano committed:
Update on "Add randomness case to the autograd notes"


I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

[ghstack-poisoned]

lezcano committed:
Update on "Add randomness case to the autograd notes"


I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

[ghstack-poisoned]

lezcano committed:
Update on "Add randomness case to the autograd notes"


I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

[ghstack-poisoned]

lezcano committed:
Update on "Add randomness case to the autograd notes"


I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

[ghstack-poisoned]

soulitzer reviewed (approved):
Looks mostly good! Have some small comments. Thanks for fixing up the sphinx.

Lezcano commented on docs/source/notes/autograd.rst:
```diff
@@ -93,15 +93,17 @@ Gradients for non-differentiable functions
+#. If the function is concave (at least locally), use the super-gradient of minimum norm (consider `-f(x)` and apply the previous point).
+#. If the function is defined, define the gradient at the current point by continuity (note that ``inf`` is possible here, for example for ``sqrt(0)``). If multiple values are possible, pick one arbitrarily.
+#. If the function is not defined (``sqrt(-1)``, ``log(-1)`` or most functions when the input is ``NaN``, for example) then the value used as the gradient is arbitrary (we might also raise an error but that is not guaranteed). Most functions will use ``NaN`` as the gradient, but for performance reasons, some functions will use other values (``log(-1)``, for example).
+#. If the function is not a deterministic mapping (i.e. it is not a `mathematical function`_), it will be marked as non-differentiable, and it will error out in the backward if used on tensors that require grad outside of a ``no_grad`` environment.
```
I didn't find a way to put the math function def before the points in a way that the text actually flows naturally, so leaving as it is for now.

I'll correct the second observation, thanks!

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -93,15 +93,17 @@ Gradients for non-differentiable functions
-#. If the function is concave (at least locally), use the super-gradient with minimum norm (using a similar argument as above).
-#. If the function is defined, define the gradient at the current point by continuity (note that :math:`inf` is possible here, for example, :math:`sqrt(0)`). If multiple values are possible, pick one arbitrarily.
-#. If the function is not defined (:math:`\sqrt(-1)`, :math:`\log(-1)` or most functions when the input is :math:`nan` for example) then the value used as the gradient is arbitrary (we might also raise an error but that is not guaranteed). Most functions will use :math:`nan` as the gradient, but for performance reasons, some functions will use non-:math:`nan` values (:math:`\log(-1)` for example) ...
+#. If the function is convex (at least locally), use the sub-gradient of minimum norm (it the steepest descent direction).
```
nit: it is

Also why remove the reference?

Lezcano replied:
I thought referencing an exercise in a random book was not too cool

if you think it's necessary, I can look up a proper reference, but also we never give references for anything else, so that reference looked a bit out of context

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -93,15 +93,17 @@ Gradients for non-differentiable functions
-#. If the function is not defined (:math:`\sqrt(-1)`, :math:`\log(-1)` or most functions when the input is :math:`nan` for example) then the value used as the gradient is arbitrary (we might also raise an error but that is not guaranteed). Most functions will use :math:`nan` as the gradient, but for performance reasons, some functions will use non-:math:`nan` values (:math:`\log(-1)` for example) ...
+#. If the function is convex (at least locally), use the sub-gradient of minimum norm (it the steepest descent direction).
+#. If the function is concave (at least locally), use the super-gradient of minimum norm (consider `-f(x)` and apply the previous point).
+#. If the function is defined, define the gradient at the current point by continuity (note that ``inf`` is possible here, for example for ``sqrt(0)``). If multiple values are possible, pick one arbitrarily.
```
not using `math` is to improve the formatting inline?

Lezcano replied:
That's what we do in `linalg` at least. The idea is that LaTeX is a bit disruptive when reading, as it changes the whole notation. We use LaTeX whenever we cannot avoid it.

Here you can even argue that you really want code formatting, as you're talking about `inf` and `sqrt(0)` which are objects from the PyTorch world.

albanD reviewed (commented):
Nice update!

lezcano committed:
Update on "Add randomness case to the autograd notes"


I also took this chance to clean a bit the sphinx formatting and
reworded a few minor things.

[ghstack-poisoned]

Lezcano commented:
@albanD this is ready for another review

albanD reviewed (approved):
SGTM
